cf-custom-resources/lib/dns-delegation.js

```diff
@@ -0,0 +1,250 @@
+            LogicalResourceId: event.LogicalResourceId,
+            Data: responseData
+        });
+
+        const parsedUrl = new URL(event.ResponseURL || defaultResponseURL);
+        const options = {
+            hostname: parsedUrl.hostname,
+            port: 443,
+            path: parsedUrl.pathname + parsedUrl.search,
+            method: 'PUT',
+            headers: {
+                'Content-Type': '',
+                'Content-Length': responseBody.length
+            }
+        };
+
+        https.request(options)
+            .on('error', reject)
+            .on('response', res => {
+                res.resume();
+                if (res.statusCode >= 400) {
+                    reject(new Error(`Server returned error ${res.statusCode}: ${res.statusMessage}`));
+                } else {
+                    resolve();
+                }
+            })
+            .end(responseBody, 'utf8');
+    });
+};
+
+/**
+ * Creates a NS recordset in the domain's hosted zone using the rootDnsRole for the subDomain.
+ * This essentially delegates authority for subdomain to the subdomain's hostedzone.
+ *
+ * The rootDnsRole has to have access to the hostedzone for domainName.
+ *
+ * @param {string} requestId the CloudFormation request ID
+ * @param {string} domainName the DNS name to add to the subDomain to (ecs-cli.aws).
+ * @param {string} subDomain the full subdomain to add to the domain above (test.ecs-cli.aws).
+ * @param {string[]} nameServers the subdomain nameservers to add to the domain's hostedzone.
+ * @param {string} rootDnsRole the IAM role ARN that can manage domainName
+ * @returns {string} the created subdomain
+ */
+const createSubdomainInRoot = async function (requestId, domainName, subDomain, nameServers, rootDnsRole) {
+    const route53 = new aws.Route53({
+        credentials: new aws.ChainableTemporaryCredentials({
+            RoleArn: rootDnsRole,
+        })
+    })
+
+    const hostedZones = await route53.listHostedZonesByName({
+        DNSName: domainName
+    }).promise();
+
+    if (!hostedZones.HostedZones || hostedZones.HostedZones.length == 0) {
+        throw new Error(`Couldn't find any hostedzones with DNS name ${domainName}. Request ${requestId}`)
+    }
+
+    const domainHostedZone = hostedZones.HostedZones[0];
+
+    // HostedZoneIDs are of the form /hostedzone/1234455, but the actual
+    // ID is after the last slash.
+    const hostedZoneId = domainHostedZone.Id.split('/').pop();
+
+
+    const changeBatch = await route53.changeResourceRecordSets({
+        ChangeBatch: {
+            Changes: [recordChangeAction('UPSERT', subDomain, "NS", nameServers.map(ns => {
+                return {
+                    Value: ns
+                }
+            }))]
+        },
+        HostedZoneId: hostedZoneId
+    }).promise();
+
+    console.log(`Created recordset in hostedzone ${hostedZoneId} for ${subDomain}`);
+
+    await waitForRecordSetChange(route53, changeBatch.ChangeInfo.Id);
```

Comment:
Should we reject if this promise fails? (same for delete)

Reply:
await will throw if the promise is rejected.